### PR TITLE
feat(release): add support for sparse registry URLs

### LINKED ISSUE #831: Can't publish to private registry (works with cargo publish)
## Bug description

Can't publish to private registry (works with cargo publish)

## Expected behavior

I would expect the crates to be published like they are with `cargo publish` or `cargo release`

## Screenshots

Screenshot 2023-06-21 at 6 22 59 PM

## Environment

- OS: macOS 13.3
- release-plz version: release-plz 0.3.14

### PR DESCRIPTION
**Pull request by Kelwing**
Adds support for sparse registry URLs, fixes #831.

### CODE REVIEW

File: crates/release_plz_core/src/cargo.rs [hunk 1]
```diff
@@ -68,6 +79,17 @@ pub fn is_published(index: &mut Index, package: &Package) -> anyhow::Result<bool
...
+pub async fn is_published_sparse(index: &SparseIndex, package: &Package) -> anyhow::Result<bool> {
+    let url = index.crate_url(&package.name);
+
+    match url {
+        // sparse protocol can fetch the manifest for a single crate, so we can
+        // hit the registry directly
+        Some(u) => Ok(reqwest::get(u).await?.status().is_success()),
+        None => Ok(false),
+    }
+}
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/cargo.rs:91]: in this function, shouldn't we verify the package version, too?

**Comment by MarcoIeni** [comment]
thank you so much. Looks very good! I just left a comment

File: crates/release_plz_core/src/cargo.rs (hunk 1 above)
Comment (Kelwing) [reply crates/release_plz_core/src/cargo.rs:84]: You're correct.  I just added that.

File: crates/release_plz_core/Cargo.toml [hunk 2]
```diff
@@ -28,7 +28,7 @@ lazy_static.workspace = true
 parse-changelog.workspace = true
 rayon.workspace = true
 regex.workspace = true
-reqwest = { workspace = true, features = ["json"] }
+reqwest = { workspace = true, features = ["json", "gzip", "native-tls-alpn"] }
```
Comment (MarcoIeni) [inline crates/release_plz_core/Cargo.toml:31]: can you add a comment saying why native-tls-alpn is required?

File: Cargo.toml [hunk 3]
```diff
@@ -9,20 +9,21 @@ chrono = { version = "0.4.26", default-features = false }
...
 git-url-parse = "0.4.4"
+http = "0.2"
 ignore = "0.4.20"
 lazy_static = "1.4.0"
 once_cell = "1.18.0"
 parse-changelog = { version = "0.6.1", default-features = false }
 rayon = "1.7.0"
 regex = "1.9.1"
-reqwest = "0.11.18"
+reqwest = { version = "0.11.18" }
```
Comment (MarcoIeni) [inline Cargo.toml:26]:
```suggestion
reqwest = "0.11.18"
```

File: crates/release_plz_core/src/cargo.rs [hunk 4]
```diff
@@ -68,6 +79,10 @@ pub fn is_published(index: &mut Index, package: &Package) -> anyhow::Result<bool
     Ok(is_in_cache(index, package))
 }
 
+pub async fn is_published_sparse(index: &SparseIndex, package: &Package) -> anyhow::Result<bool> {
+    is_in_cache_sparse(index, package).await
+}
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/cargo.rs:84]: This function only calls another function. Should we remove it? For example, we can directly call `is_in_cache_sparse`

**Comment by MarcoIeni** [comment]
In https://github.com/MarcoIeni/release-plz/commit/aece670abababc7ef67c7d5b3fa560ddb5dbc89e I removed some duplicate code. If you like these changes, can you apply them to your branch?
Sorry for the manual task, but I didn't know a better way to propose these changes to your pr :(

**Comment by MarcoIeni** [comment]
the code looks good! I will test this and it if works, I will merge it 👍

**Comment by MarcoIeni** [comment]
tested by adding 

```
[registries.crates-io-http]
index = "sparse+https://index.crates.io/"
```

to `.cargo/config.toml` and running `release-plz release --registry crates-io-http`.
It worked! Thanks so much for this contribution. 🙏

**Pull request merged**